src/Plutarch/Extra/Precompile.hs

```diff
@@ -0,0 +1,240 @@
+    CompiledTerm (a :--> b) ->
+    (forall (s :: S). Term s a) ->
+    CompiledTerm b
+applyCompiledTerm (CompiledTerm sf) a =
+    CompiledTerm $ applyDScript sf (mustEvalD $ mustCompileD a)
+
+{- | Apply a 'CompiledTerm' to a closed Plutarch 'Term'.
+
+ Does NOT evaluate the argument before applying. Using this seems to save very
+ little overhead, not worth it for efficiency. Only use it to make argument
+ evaluation count for benchmarking.
+-}
+applyCompiledTerm' ::
+    forall (a :: S -> Type) (b :: S -> Type).
+    CompiledTerm (a :--> b) ->
+    (forall (s :: S). Term s a) ->
+    CompiledTerm b
+applyCompiledTerm' (CompiledTerm sf) a =
+    CompiledTerm $ applyDScript sf (mustCompileD a)
+
+{- | Apply a 'CompiledTerm' to a 'CompiledTerm'.
+
+ Evaluates the argument before applying. You want this for benchmarking the
+ compiled function. Helps to avoid tainting the measurement by input
+ conversions.
+-}
+applyCompiledTerm2 ::
+    forall (a :: S -> Type) (b :: S -> Type).
+    CompiledTerm (a :--> b) ->
+    CompiledTerm a ->
+    CompiledTerm b
+applyCompiledTerm2 (CompiledTerm sf) (CompiledTerm sa) =
+    CompiledTerm $ applyDScript sf (mustEvalD sa)
+
+{- | Apply a 'CompiledTerm' to a 'CompiledTerm'.
+
+ Does NOT evaluate the argument before applying. Using this seems to save very
+ little overhead, not worth it for efficiency. Only use it to make argument
+ evaluation count for benchmarking.
+-}
+applyCompiledTerm2' ::
+    forall (a :: S -> Type) (b :: S -> Type).
+    CompiledTerm (a :--> b) ->
+    CompiledTerm a ->
+    CompiledTerm b
+applyCompiledTerm2' (CompiledTerm sf) (CompiledTerm sa) =
+    CompiledTerm $ applyDScript sf sa
+
+-- | Alias for 'applyCompiledTerm'.
+(##) ::
+    forall (a :: S -> Type) (b :: S -> Type).
+    CompiledTerm (a :--> b) ->
+    (forall (s :: S). Term s a) ->
+    CompiledTerm b
+(##) = applyCompiledTerm
+
+infixl 8 ##
+
+-- | Alias for 'applyCompiledTerm\''.
+(##~) ::
+    forall (a :: S -> Type) (b :: S -> Type).
+    CompiledTerm (a :--> b) ->
+    (forall (s :: S). Term s a) ->
+    CompiledTerm b
+(##~) = applyCompiledTerm'
+
+infixl 8 ##~
+
+-- | Alias for 'applyCompiledTerm2'.
+(###) ::
+    forall (a :: S -> Type) (b :: S -> Type).
+    CompiledTerm (a :--> b) ->
+    CompiledTerm a ->
+    CompiledTerm b
+(###) = applyCompiledTerm2
+
+infixl 7 ###
+
+-- | Alias for 'applyCompiledTerm2\''.
+(###~) ::
+    forall (a :: S -> Type) (b :: S -> Type).
+    CompiledTerm (a :--> b) ->
+    CompiledTerm a ->
+    CompiledTerm b
+(###~) = applyCompiledTerm2'
+
+infixl 7 ###~
+
+--  Copied and adapted the stuff below from 'Plutarch.Lift'. Including
+--  'LiftError', because the data constructors aren't exported there. Also
+--  added trace messages in the exceptions.
+
+-- | Error during script evaluation.
+data LiftError
+    = LiftError_EvalError EvalError [Text]
+    | LiftError_KnownTypeError KnownTypeError
+    | LiftError_FromRepr
+    deriving stock (Eq)
```

Comment:
The underscores in the names here aren't really idiomatic Haskell. Also, is something stopping this being a `Show` instance?

Reply:
This is lifted directly from [`Plutarch.Lift`](https://hoogle.nix.dance/file/nix/store/7r3zwxqgrvyd38z8myq5n9w7hzgnvj1m-plutarch-lib-plutarch-1.1.0-haddock-doc/share/doc/plutarch/html/src/Plutarch.Lift.html#LiftError). I've just asked in the discord if we can export these data constructors to simplify some of this code and the following functions.

Reply:
@cjay may have more comments here

Reply:
Yeah I just tried to make minimal changes to the original code.

Reply:
Btw, the original now includes a `LiftError_CompilationError`. We don't even want that, because the Script is already compiled.

Reply:
I've gotten the OK to make a PR to export these

Reply:
> Is something stopping this being a `Show` instance?

The wrapped PlutusCore type doesn't have a show instance.

Reply:
https://github.com/Plutonomicon/plutarch-plutus/pull/535